### PR TITLE
Enable Falcon-mamba

### PR DESCRIPTION
**Pull request by yuanwu2017**
# What does this PR do?
This PR is for enabling the falcon-mamba model.
Gaudi2:
` python run_generation.py --model_name_or_path tiiuae/falcon-mamba-7b  --use_hpu_graphs --use_kv_cache --max_new_tokens 100 --do_sample --batch_size 1 --prompt "How are you?" --bf16
`
Input/outputs:
input 1: ('How are you?',)
output 1: ('How are you?\nHow are you doing?\nHow are things?\nHow are they going?\nWhen someone answers "fine," they mostly or always mean "good." When someone answers "not well," they always mean "trouble."\nThat is why in introductory e-mails you will notice that some say "Hope you\'re in having a good summer" and others say "how are you doing?" and which would it be if your answer was "Its',)

Stats:
---------------------------------------------------------------------------------
Input tokens
Throughput (including tokenization) = 47.1464839567739 tokens/second
Memory allocated                    = 13.6 GB
Max memory allocated                = 13.6 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 40.49411549489014 seconds
---------------------------------------------------------------------------------

A100:
`python run_generation.py --model_name_or_path tiiuae/falcon-mamba-7b  --use_kv_cache --max_new_tokens 100 --do_sample --batch_size 1 --prompt "How are you?"  --device cuda`

Input/outputs:
input 1: ('How are you?',)
output 1: ("How are you?', 'Thanks for stopping by!', 'Hello friend!']\n\nIn both examples, the methods are not used to select items, but rather to browse through either an array of items describing its state or an object containing key-value pairs, representing an initial setup of the configuration. The `.each()` method allows for seamless and efficient traversal of the documentation, presenting the various states that the items or configuration options may have in a structured format, rather than simply printing their",)

Stats:
-----------------------------------------------------------------------------------
Input tokens
Throughput (including tokenization) = 24.131639356423253 tokens/second
-----------------------------------------------------------------------------------

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by yuanwu2017** [comment]
@regisss @libinta @mandy-li  Please help to review.

File: optimum/habana/transformers/models/falcon_mamba/modeling_falcon_mamba.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    if use_cache:
+        # `cache_position` should have been initialized in `generate`
+        if cache_position is None:
+            raise ValueError(
+                "`cache_position` should not be None as it should have been initialized in "
+                "`model.generate`, you are responsible for passing in a valid `cache_position` if "
+                "you are calling `prepare_inputs_for_generation` directly with `use_cache=True`"
+            )
+        if cache_position[0] > 0:
+            # input_ids = input_ids[:, -1].unsqueeze(-1)
```
Comment (jiminha) [inline optimum/habana/transformers/models/falcon_mamba/modeling_falcon_mamba.py:54]: Is this for throughput, or for avoiding recompilation?

**Comment by jiminha** [comment]
Also, what other configurations have been tested besides the example above and compare the result?
Like input/output 1K, 4K, 16K and etc? and max_batch size?

File: optimum/habana/transformers/models/falcon_mamba/modeling_falcon_mamba.py (hunk 1 above)
Comment (yuanwu2017) [reply optimum/habana/transformers/models/falcon_mamba/modeling_falcon_mamba.py:54]: For avoiding recompilation, it causes that the function of _hpu_C.get_view_hash gets the different hash values for same shape tensor. I will fill an issue ticket for habana pytorch team.  

logs:
```
obj=tensor([[4]], device='hpu:0')
obj.shape=torch.Size([1, 1])
_hpu_C.get_view_hash(obj)=16266269407708533847

obj=tensor([[204]], device='hpu:0')
obj.shape=torch.Size([1, 1])
_hpu_C.get_view_hash(obj)=16266925452406130564  
```

**Comment by yuanwu2017** [comment]
I tried the input = 1024,  it token about 1621 seconds to compiling the graphs. Currently we have no good way to solve it.
image
This problem is mainly caused by our lack of optimized ops.  The following code takes a very long time to generate a graph, and it is directly related to the input length.
https://github.com/huggingface/transformers/blob/v4.45.2/src/transformers/models/falcon_mamba/modeling_falcon_mamba.py#L363
```
  for i in range(seq_len):
                ssm_state = (
                    discrete_A[:, :, i, :] * ssm_state + deltaB_u[:, :, i, :]
                )  # [batch, intermediate_size, ssm_state]
                scan_output = torch.matmul(
                    ssm_state.to(dtype), C[:, i, :].unsqueeze(-1)
                )  # [batch, intermediate_size, 1]
                scan_outputs.append(scan_output[:, :, 0])
            scan_output = torch.stack(scan_outputs, dim=-1)  # [batch, intermediate_size, seq_len]
            scan_output = scan_output + (hidden_states * self.D[None, :, None])
            scan_output = scan_output * self.act(gate)
```
For cuda device, they have optimized ops.
https://github.com/huggingface/transformers/blob/v4.45.2/src/transformers/models/falcon_mamba/modeling_falcon_mamba.py#L248
https://github.com/state-spaces/mamba/blob/442fab4b1fd5226c1b5939b37d91ede430b5d1ae/setup.py#L238

**Comment by yuanwu2017** [comment]
@libinta @mandy-li @jiminha Please help to review the patch

**Comment by jiminha** [comment]
@ssarkar2   Do you have any suggestion on this? It seems longer sequence the compile time is huge on this section.

@zzhang37 Could you also check on this? Do we have similar case on mamba? 


> This problem is mainly caused by our lack of optimized ops. The following code takes a very long time to generate a graph, and it is directly related to the input length, so output_length(max_new_tokens) will not cause warmup to be obviously too long https://github.com/huggingface/transformers/blob/v4.45.2/src/transformers/models/falcon_mamba/modeling_falcon_mamba.py#L363
> 
> ```
>   for i in range(seq_len):
>                 ssm_state = (
>                     discrete_A[:, :, i, :] * ssm_state + deltaB_u[:, :, i, :]
>                 )  # [batch, intermediate_size, ssm_state]
>                 scan_output = torch.matmul(
>                     ssm_state.to(dtype), C[:, i, :].unsqueeze(-1)
>                 )  # [batch, intermediate_size, 1]
>                 scan_outputs.append(scan_output[:, :, 0])
>             scan_output = torch.stack(scan_outputs, dim=-1)  # [batch, intermediate_size, seq_len]
>             scan_output = scan_output + (hidden_states * self.D[None, :, None])
>             scan_output = scan_output * self.act(gate)
> ```
> 
> For cuda device, they have optimized ops. https://github.com/huggingface/transformers/blob/v4.45.2/src/transformers/models/falcon_mamba/modeling_falcon_mamba.py#L248 https://github.com/state-spaces/mamba/blob/442fab4b1fd5226c1b5939b37d91ede430b5d1ae/setup.py#L238

**Comment by jiminha** [comment]
@yuanwu2017 can you open HS ticket and list all the variation that you tested 


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jiminha/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jiminha/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">


batch_size | input len | output len | compile time | first token | throughput | ref throughput | memory
-- | -- | -- | -- | -- | -- | -- | --
1 | 128 | 128 |   |   |   |   |  
max_batch or best batch | 128 | 128 |   |   |   |   |  
  | 1024 | 1024 |   |   |   |   |  
  | 2048 | 2048 |   |   |   |   |  
  | 128 | 2048 |   |   |   |   |  
  | 2048 | 128 |   |   |   |   |  
  |   | 4k?   16? |   |   |   |   |  
  |   | large   sequence length? |   |   |   |   |  



</body>

</html>

**Review by jiminha — APPROVED** [review]

**Comment by jiminha** [comment]
> @jiminha Please review.
Please check HS ticket, I think some numbers don’t look right for certain configurations. I believe there is a lot of room for further optimization. We will take the PR as it is for now.

**Comment by regisss** [comment]
@jiminha I guess we don't want to add this model to the table of validated models in the README yet, right? We did the same for Mamba: it is supported but since the performance is not good enough yet, we decided to not add it to the table.

**Comment by libinta** [comment]
@regisss  yes, we don't need to add this to the table

**Review by regisss — APPROVED** [review]
